Common/src/geometry/meshreader/CSU2BinaryMeshReaderBase.cpp

```diff
@@ -0,0 +1,470 @@
+     and allocate the memory for the send requests. */
+  int nRankSend = 0;
+  for (int i = 0; i < size; ++i) {
+    if (sendToRank[i]) ++nRankSend;
+  }
+
+  vector<SU2_MPI::Request> sendReqs(nRankSend);
+
+  /* Send the data using non-blocking sends. */
+  nRankSend = 0;
+  for (int i = 0; i < size; ++i) {
+    if (sendToRank[i]) {
+      SU2_MPI::Isend(send_buf[i].data(), send_buf[i].size(), MPI_UNSIGNED_LONG, i, i, SU2_MPI::GetComm(),
+                     &sendReqs[nRankSend]);
+      ++nRankSend;
+    }
+  }
+
+  /* Define the receive buffers and receive the messages. */
+  std::vector<std::vector<unsigned long>> recv_buf;
+  recv_buf.resize(size);
+
+  for (int i = 0; i < nRankRecv; ++i) {
+    SU2_MPI::Status status;
+    SU2_MPI::Probe(MPI_ANY_SOURCE, rank, SU2_MPI::GetComm(), &status);
+    int rankRecv = status.MPI_SOURCE;
+
+    int sizeMess;
+    SU2_MPI::Get_count(&status, MPI_UNSIGNED_LONG, &sizeMess);
+    recv_buf[rankRecv].resize(sizeMess);
+    SU2_MPI::Recv(recv_buf[rankRecv].data(), sizeMess, MPI_UNSIGNED_LONG, rankRecv, rank, SU2_MPI::GetComm(), &status);
+  }
+
+  /* Complete the non-blocking sends and release the memory of the send buffers. */
+  SU2_MPI::Waitall(nRankSend, sendReqs.data(), MPI_STATUSES_IGNORE);
+  for (int i = 0; i < size; ++i) {
+    if (sendToRank[i]) vector<unsigned long>().swap(send_buf[i]);
+  }
+
+  /* Synchronize the MPI ranks, because wild cards have been used. */
+  SU2_MPI::Barrier(SU2_MPI::GetComm());
+
+  /*--- Store the information in the receive buffers in the offset and conn_buff
+        vectors, such that it is consistent with the information without MPI.
+        Release the memory of the receive buffers afterwards. ---*/
+  offset.push_back(0);
+  for (int i = 0; i < size; ++i) {
+    if (recv_buf[i].size() > 0) {
+      size_t ind = 0;
+      while (ind < recv_buf[i].size()) {
+        auto n_items = recv_buf[i][ind++];
+        offset.push_back(offset.back() + n_items);
+        for (unsigned long j = 0; j < n_items; ++j, ++ind) conn_buff.push_back(recv_buf[i][ind]);
+      }
+      vector<unsigned long>().swap(recv_buf[i]);
+    }
+  }
+
+#endif
+
+  /*--- Extract the connectivity data from conn_buf and store the data
+        in the appropriate member variables. ---*/
+  numberOfLocalElements = offset.size() - 1;
+  array<unsigned long, N_POINTS_HEXAHEDRON> connectivity{};
+
+  for (unsigned long i = 0; i < numberOfLocalElements; ++i) {
+    auto ind = offset[i];
+    auto size_this_elem = static_cast<unsigned short>(offset[i + 1] - offset[i]);
+    auto VTK_Type = conn_buff[ind++];
+    const auto nPointsElem = nPointsOfElementType(static_cast<unsigned short>(VTK_Type));
+    if (size_this_elem < (nPointsElem + 2)) SU2_MPI::Error("Not enough items in volume connectivity", CURRENT_FUNCTION);
+
+    for (unsigned short j = 0; j < nPointsElem; ++j, ++ind) connectivity[j] = conn_buff[ind];
+    auto GlobalIndex = conn_buff[ind];
```

Comment:
Is the global index used? I have the feeling we always end up ignoring the file and basing things on the order of elements in the file